File: ip/lib/src/golden/vision.dart
// Vision Transformer encoder golden (CLIP / SigLIP style), pure fp64 reference.
// Pipeline: patch-embed (a conv lowered to im2col + matmul) + optional class
// token + learned position embeddings, then N pre-norm ViT blocks (LayerNorm +
// bidirectional multi-head attention + GELU MLP), then a final LayerNorm. Same
// matmul datapath as the LLM. The differences are LayerNorm (not RMSNorm), GELU
// (not SiLU), learned positions, non-causal attention, and the patch conv.
import 'dart:math' as math;
import 'dart:typed_data';

import '../ir/model_graph.dart';
import '../weights/weight_store.dart';
import 'ops.dart';

/// One ViT block's weights: pre-attention LayerNorm, q/k/v/o projections (with
/// biases), pre-MLP LayerNorm, and the two MLP projections (with biases).
class VisionBlockWeights {
  final Float64List ln1Gamma;
  final Float64List ln1Beta;
  final Float64List qProj;
  final Float64List qBias;
  final Float64List kProj;
  final Float64List kBias;
  final Float64List vProj;
  final Float64List vBias;
  final Float64List oProj;
  final Float64List oBias;
  final Float64List ln2Gamma;
  final Float64List ln2Beta;
  final Float64List fc1;
  final Float64List fc1Bias;
  final Float64List fc2;
  final Float64List fc2Bias;

  const VisionBlockWeights({
    required this.ln1Gamma,
    required this.ln1Beta,
    required this.qProj,
    required this.qBias,
    required this.kProj,
    required this.kBias,
    required this.vProj,
    required this.vBias,
    required this.oProj,
    required this.oBias,
    required this.ln2Gamma,
    required this.ln2Beta,
    required this.fc1,
    required this.fc1Bias,
    required this.fc2,
    required this.fc2Bias,
  });
}

/// A full vision tower's decoded weights plus dimensions.
class VisionWeights {
  /// Patch-embedding conv, flattened to `[hidden x (numChannels*patch*patch)]`
  /// (out-channel major, then in-channel, then kernel row/col), and its bias.
  final Float64List patchEmbed;
  final Float64List patchEmbedBias;

  /// Learned class token `[hidden]`, prepended when [hasClassToken].
  final Float64List? classToken;

  /// Learned position embeddings `[seqLen x hidden]`.
  final Float64List posEmbed;

  /// Optional pre-encoder LayerNorm (CLIP's `pre_layrnorm`. SigLIP has none).
  final Float64List? preLnGamma;
  final Float64List? preLnBeta;

  final List<VisionBlockWeights> blocks;

  /// Final post-encoder LayerNorm.
  final Float64List postLnGamma;
  final Float64List postLnBeta;

  final int hidden;
  final int numHeads;
  final int headDim;
  final int intermediate;
  final int patchSize;
  final int numChannels;
  final int imageSize;
  final double lnEps;
  final bool hasClassToken;

  const VisionWeights({
    required this.patchEmbed,
    required this.patchEmbedBias,
    required this.classToken,
    required this.posEmbed,
    required this.preLnGamma,
    required this.preLnBeta,
    required this.blocks,
    required this.postLnGamma,
    required this.postLnBeta,
    required this.hidden,
    required this.numHeads,
    required this.headDim,
    required this.intermediate,
    required this.patchSize,
    required this.numChannels,
    required this.imageSize,
    required this.lnEps,
    required this.hasClassToken,
  });

  int get patchesPerSide => imageSize ~/ patchSize;
  int get numPatches => patchesPerSide * patchesPerSide;
  int get seqLen => numPatches + (hasClassToken ? 1 : 0);
}

/// Resolves and decodes a vision tower's weights from [store] using the
/// HuggingFace CLIP/SigLIP naming under [prefix] (default `vision_model`). The
/// patch-embedding conv weight (`[hidden, C, P, P]`) flattens to the
/// `[hidden, C*P*P]` matrix [patchify] expects. Class token and pre-LayerNorm
/// are loaded only when present (CLIP has both. SigLIP has neither).
VisionWeights bindVision(
  VisionTowerSpec s,
  WeightStore store, {
  String prefix = 'vision_model',
}) {
  Float64List get(String name) {
    if (!store.contains(name)) {
      throw ArgumentError('missing vision tensor: $name');
    }
    return store.get(name).toFloat64List();
  }

  Float64List? getOpt(String name) =>
      store.contains(name) ? store.get(name).toFloat64List() : null;

  final blocks = <VisionBlockWeights>[
    for (var i = 0; i < s.numLayers; i++)
      VisionBlockWeights(
        ln1Gamma: get('$prefix.encoder.layers.$i.layer_norm1.weight'),
        ln1Beta: get('$prefix.encoder.layers.$i.layer_norm1.bias'),
        qProj: get('$prefix.encoder.layers.$i.self_attn.q_proj.weight'),
        qBias: get('$prefix.encoder.layers.$i.self_attn.q_proj.bias'),
        kProj: get('$prefix.encoder.layers.$i.self_attn.k_proj.weight'),
        kBias: get('$prefix.encoder.layers.$i.self_attn.k_proj.bias'),
        vProj: get('$prefix.encoder.layers.$i.self_attn.v_proj.weight'),
        vBias: get('$prefix.encoder.layers.$i.self_attn.v_proj.bias'),
        oProj: get('$prefix.encoder.layers.$i.self_attn.out_proj.weight'),
        oBias: get('$prefix.encoder.layers.$i.self_attn.out_proj.bias'),
        ln2Gamma: get('$prefix.encoder.layers.$i.layer_norm2.weight'),
        ln2Beta: get('$prefix.encoder.layers.$i.layer_norm2.bias'),
        fc1: get('$prefix.encoder.layers.$i.mlp.fc1.weight'),
        fc1Bias: get('$prefix.encoder.layers.$i.mlp.fc1.bias'),
        fc2: get('$prefix.encoder.layers.$i.mlp.fc2.weight'),
        fc2Bias: get('$prefix.encoder.layers.$i.mlp.fc2.bias'),
      ),
  ];

  return VisionWeights(
    patchEmbed: get('$prefix.embeddings.patch_embedding.weight'),
    patchEmbedBias:
        getOpt('$prefix.embeddings.patch_embedding.bias') ??
        Float64List(s.hiddenSize),
    classToken: s.hasClassToken
        ? get('$prefix.embeddings.class_embedding')
        : null,
    posEmbed: get('$prefix.embeddings.position_embedding.weight'),
    preLnGamma: getOpt('$prefix.pre_layrnorm.weight'),
    preLnBeta: getOpt('$prefix.pre_layrnorm.bias'),
    blocks: blocks,
    postLnGamma: get('$prefix.post_layernorm.weight'),
    postLnBeta: get('$prefix.post_layernorm.bias'),
    hidden: s.hiddenSize,
    numHeads: s.numHeads,
    headDim: s.headDim,
    intermediate: s.intermediateSize,
    patchSize: s.patchSize,
    numChannels: s.numChannels,
    imageSize: s.imageSize,
    lnEps: s.layerNormEps,
    hasClassToken: s.hasClassToken,
  );
}

/// Splits a channels-first image `[numChannels x imageSize x imageSize]` into
/// non-overlapping `patchSize` patches, each flattened to length
/// `numChannels*patch*patch` in (channel, row, col) order to match the conv
/// weight layout. Returns `numPatches` patch vectors in row-major patch order.
List<Float64List> patchify(Float64List pixels, VisionWeights w) {
  final c = w.numChannels, s = w.imageSize, p = w.patchSize;
  final pps = s ~/ p;
  final expected = c * s * s;
  if (pixels.length != expected) {
    throw ArgumentError('pixels.length ${pixels.length} != $expected (C*S*S)');
  }
  final patchLen = c * p * p;
  final out = <Float64List>[];
  for (var py = 0; py < pps; py++) {
    for (var px = 0; px < pps; px++) {
      final vec = Float64List(patchLen);
      var idx = 0;
      for (var ch = 0; ch < c; ch++) {
        for (var dy = 0; dy < p; dy++) {
          for (var dx = 0; dx < p; dx++) {
            vec[idx++] = pixels[ch * s * s + (py * p + dy) * s + (px * p + dx)];
          }
        }
      }
      out.add(vec);
    }
  }
  return out;
}

/// Patch-embed + class token + position embeddings. Returns `seqLen` rows.
List<Float64List> visionEmbed(Float64List pixels, VisionWeights w) {
  final patchLen = w.numChannels * w.patchSize * w.patchSize;
  final patches = patchify(pixels, w);
  final rows = <Float64List>[];
  if (w.hasClassToken) {
    rows.add(Float64List.fromList(w.classToken!));
  }
  for (final patch in patches) {
    rows.add(
      linear(w.patchEmbed, w.hidden, patchLen, patch, bias: w.patchEmbedBias),
    );
  }
  // Add learned position embeddings (row i gets posEmbed[i]).
  for (var i = 0; i < rows.length; i++) {
    for (var d = 0; d < w.hidden; d++) {
      rows[i][d] += w.posEmbed[i * w.hidden + d];
    }
  }
  return rows;
}

/// Bidirectional (non-causal) multi-head attention: every position attends to
/// every position. `q`/`k`/`v` are `seqLen` rows of `numHeads*headDim`.
List<Float64List> bidirectionalAttention(
  List<Float64List> q,
  List<Float64List> k,
  List<Float64List> v,
  int numHeads,
  int headDim,
) {
  final t = q.length;
  final dim = numHeads * headDim;
  final scale = 1.0 / math.sqrt(headDim.toDouble());
  final out = List<Float64List>.generate(t, (_) => Float64List(dim));
  for (var h = 0; h < numHeads; h++) {
    final off = h * headDim;
    for (var i = 0; i < t; i++) {
      final scores = Float64List(t);
      for (var j = 0; j < t; j++) {
        var dot = 0.0;
        for (var d = 0; d < headDim; d++) {
          dot += q[i][off + d] * k[j][off + d];
        }
        scores[j] = dot * scale;
      }
      final sm = softmax(scores);
      for (var j = 0; j < t; j++) {
        final weight = sm[j];
        for (var d = 0; d < headDim; d++) {
          out[i][off + d] += weight * v[j][off + d];
        }
      }
    }
  }
  return out;
}

/// One pre-norm ViT block: `x + attn(LN1(x))` then `h + mlp(LN2(h))`.
List<Float64List> visionBlock(
  List<Float64List> x,
  VisionBlockWeights bw,
  VisionWeights w,
) {
  final t = x.length;
  final q = <Float64List>[];
  final k = <Float64List>[];
  final v = <Float64List>[];
  for (var i = 0; i < t; i++) {
    final n1 = layerNorm(x[i], bw.ln1Gamma, bw.ln1Beta, w.lnEps);
    q.add(linear(bw.qProj, w.hidden, w.hidden, n1, bias: bw.qBias));
    k.add(linear(bw.kProj, w.hidden, w.hidden, n1, bias: bw.kBias));
    v.add(linear(bw.vProj, w.hidden, w.hidden, n1, bias: bw.vBias));
  }
  final attn = bidirectionalAttention(q, k, v, w.numHeads, w.headDim);

  final h1 = <Float64List>[];
  for (var i = 0; i < t; i++) {
    final o = linear(bw.oProj, w.hidden, w.hidden, attn[i], bias: bw.oBias);
    final r = Float64List(w.hidden);
    for (var d = 0; d < w.hidden; d++) {
      r[d] = x[i][d] + o[d];
    }
    h1.add(r);
  }

  final out = <Float64List>[];
  for (var i = 0; i < t; i++) {
    final n2 = layerNorm(h1[i], bw.ln2Gamma, bw.ln2Beta, w.lnEps);
    final f1 = linear(bw.fc1, w.intermediate, w.hidden, n2, bias: bw.fc1Bias);
    final act = gelu(f1);
    final f2 = linear(bw.fc2, w.hidden, w.intermediate, act, bias: bw.fc2Bias);
    final r = Float64List(w.hidden);
    for (var d = 0; d < w.hidden; d++) {
      r[d] = h1[i][d] + f2[d];
    }
    out.add(r);
  }
  return out;
}

/// Encodes a preprocessed image (`[numChannels x imageSize x imageSize]`,
/// already resized + normalized) into `seqLen` patch embeddings of `hidden`.
List<Float64List> encodeImage(Float64List pixels, VisionWeights w) {
  var x = visionEmbed(pixels, w);
  if (w.preLnGamma != null) {
    x = [
      for (final row in x) layerNorm(row, w.preLnGamma!, w.preLnBeta!, w.lnEps),
    ];
  }
  for (final bw in w.blocks) {
    x = visionBlock(x, bw, w);
  }
  return [
    for (final row in x) layerNorm(row, w.postLnGamma, w.postLnBeta, w.lnEps),
  ];
}

File: runtime/lib/loom/vision.zig
//! Device Vision Transformer encoder + multimodal projector. Mirrors the golden
//! reference golden/vision.dart + golden/projector.dart: patch-embed (im2col +
//! int4 matmul) + optional class token + learned position embeddings, then N
//! pre-norm ViT blocks (LayerNorm + bidirectional attention + GELU MLP), a final
//! LayerNorm, and the projector into the LLM's token-embedding space. Every
//! matmul runs on the device (W4A8); LayerNorm/GELU/attention are host glue.
const std = @import("std");
const Io = std.Io;

const model = @import("model.zig");
const linear = @import("linear.zig");
const ops = @import("ops.zig");
const Device = @import("device.zig").Device;

/// A vision matmul: column-tiled when the matrix is wider than the accelerator's
/// capacity (its weights are then stored col-block-contiguous), else single-shot.
/// `scratch` is col-tiling row scratch (>= mat.rows).
fn vmm(io: Io, dev: Device, m: model.Manifest, pt: usize, mat: model.Matrix, x: []const f32, out: []f32, scratch: []f32) !void {
    if (m.max_cols > 0 and mat.cols > m.max_cols) {
        try linear.linearColTiled(io, dev, m.csr_base, m.stores(), pt, mat, m.max_cols, x, out, scratch);
    } else {
        try linear.linear(io, dev, m.csr_base, m.stores(), pt, mat, x, out);
    }
}

/// Looks up `vision.layers.<li>.<kind>` (the returned Matrix does not alias
/// `buf`, so `buf` may be reused after).
fn vmat(cfg: model.Config, buf: []u8, li: usize, kind: []const u8) ?model.Matrix {
    const name = std.fmt.bufPrint(buf, "vision.layers.{d}.{s}", .{ li, kind }) catch return null;
    return cfg.matrix(name);
}

/// Full bidirectional (non-causal) multi-head attention over `seq` rows of
/// `q`/`k`/`vv` (`seq * nheads*hd` row-major), into `out`. `scores` is `seq`.
fn bidirAttention(q: []const f32, k: []const f32, vv: []const f32, seq: usize, nheads: usize, hd: usize, scores: []f32, out: []f32) void {
    const h = nheads * hd;
    const scale = 1.0 / @sqrt(@as(f32, @floatFromInt(hd)));
    @memset(out, 0);
    for (0..nheads) |hh| {
        const off = hh * hd;
        for (0..seq) |i| {
            for (0..seq) |j| {
                var dot: f32 = 0;
                for (0..hd) |d| dot += q[i * h + off + d] * k[j * h + off + d];
                scores[j] = dot * scale;
            }
            ops.softmaxInPlace(scores[0..seq]);
            for (0..seq) |j| {
                const w = scores[j];
                for (0..hd) |d| out[i * h + off + d] += w * vv[j * h + off + d];
            }
        }
    }
}

/// Encodes `pixels` (`num_channels*image_size*image_size`, channels-first, already
/// resized + normalized) into `seq_len` projected embeddings in the text space,
/// written row-major into `out` (`seq_len * projector.output_dim`). Allocates its
/// own scratch and frees it. Requires the model to carry a vision tower + projector.
pub fn encodeImage(gpa: std.mem.Allocator, io: Io, dev: Device, cfg: model.Config, pt: usize, pixels: []const f32, out: []f32) !void {
    const m = cfg.manifest();
    const v = m.vision orelse return error.NoVisionTower;
    const proj = m.projector orelse return error.NoProjector;
    const h: usize = v.hidden;
    const seq: usize = v.seq_len;
    const inter: usize = v.intermediate;
    const patch_len: usize = v.num_channels * v.patch_size * v.patch_size;
    const eps = v.layer_norm_eps;
    const out_dim: usize = proj.output_dim;

    // Idefics3 pixel shuffle merges scale^2 neighbouring patches before the
    // projector, so the projected token count shrinks by scale^2.
    const scale: usize = proj.scale_factor;
    const proj_tokens: usize = if (scale > 1) seq / (scale * scale) else seq;

    if (pixels.len != v.num_channels * v.image_size * v.image_size) return error.BadImageSize;
    if (out.len != proj_tokens * out_dim) return error.BadOutputSize;

    const rows = try gpa.alloc(f32, seq * h);
    defer gpa.free(rows);
    const q = try gpa.alloc(f32, seq * h);
    defer gpa.free(q);
    const k = try gpa.alloc(f32, seq * h);
    defer gpa.free(k);
    const vv = try gpa.alloc(f32, seq * h);
    defer gpa.free(vv);
    const attn = try gpa.alloc(f32, seq * h);
    defer gpa.free(attn);
    const norm = try gpa.alloc(f32, h);
    defer gpa.free(norm);
    const gamma = try gpa.alloc(f32, h);
    defer gpa.free(gamma);
    const beta = try gpa.alloc(f32, h);
    defer gpa.free(beta);
    const patch = try gpa.alloc(f32, patch_len);
    defer gpa.free(patch);
    const fc1o = try gpa.alloc(f32, inter);
    defer gpa.free(fc1o);
    const tmp = try gpa.alloc(f32, @max(h, inter));
    defer gpa.free(tmp);
    const projh = try gpa.alloc(f32, proj.hidden_dim);
    defer gpa.free(projh);
    const scores = try gpa.alloc(f32, seq);
    defer gpa.free(scores);
    // Column-tiling scratch, sized to the widest matmul's output rows.
    const cscratch = try gpa.alloc(f32, @max(inter, @max(h, out_dim)));
    defer gpa.free(cscratch);
    var name_buf: [64]u8 = undefined;

    const patch_mat = cfg.matrix("vision.patch_embed") orelse return error.MissingMatrix;

    // 1. Patch-embed. Class token (if any) is row 0; patches follow.
    const s = v.image_size;
    const p = v.patch_size;
    const pps = s / p;
    const first: usize = if (v.has_class_token) 1 else 0;
    if (v.has_class_token) cfg.readGlue(v.class_token.?, rows[0..h]);
    var pr: usize = 0;
    while (pr < pps * pps) : (pr += 1) {
        const py = pr / pps;
        const px = pr % pps;
        var idx: usize = 0;
        var ch: usize = 0;
        while (ch < v.num_channels) : (ch += 1) {
            var dy: usize = 0;
            while (dy < p) : (dy += 1) {
                var dx: usize = 0;
                while (dx < p) : (dx += 1) {
                    patch[idx] = pixels[ch * s * s + (py * p + dy) * s + (px * p + dx)];
                    idx += 1;
                }
            }
        }
        const row = rows[(first + pr) * h ..][0..h];
        try vmm(io, dev, m, pt, patch_mat, patch, row, cscratch);
        cfg.addGlue(v.patch_embed_bias, row);
    }
    // Add learned position embeddings.
    for (0..seq) |i| {
        cfg.readGlue(v.pos_embed + @as(u32, @intCast(i * h)), tmp[0..h]);
        for (rows[i * h ..][0..h], tmp[0..h]) |*r, pe| r.* += pe;
    }

    // 2. Optional pre-LayerNorm.
    if (v.pre_ln_gamma) |pg| {
        cfg.readGlue(pg, gamma);
        cfg.readGlue(v.pre_ln_beta.?, beta);
        for (0..seq) |i| {
            const r = rows[i * h ..][0..h];
            ops.layerNorm(r, gamma, beta, eps, norm);
            @memcpy(r, norm);
        }
    }

    // 3. Transformer blocks.
    for (v.layers, 0..) |lg, li| {
        const qmat = vmat(cfg, &name_buf, li, "q_proj") orelse return error.MissingMatrix;
        const kmat = vmat(cfg, &name_buf, li, "k_proj") orelse return error.MissingMatrix;
        const vmat_ = vmat(cfg, &name_buf, li, "v_proj") orelse return error.MissingMatrix;
        const omat = vmat(cfg, &name_buf, li, "out_proj") orelse return error.MissingMatrix;

        cfg.readGlue(lg.ln1_gamma, gamma);
        cfg.readGlue(lg.ln1_beta, beta);
        for (0..seq) |i| {
            const r = rows[i * h ..][0..h];
            ops.layerNorm(r, gamma, beta, eps, norm);
            try vmm(io, dev, m, pt, qmat, norm, q[i * h ..][0..h], cscratch);
            cfg.addGlue(lg.q_bias, q[i * h ..][0..h]);
            try vmm(io, dev, m, pt, kmat, norm, k[i * h ..][0..h], cscratch);
            cfg.addGlue(lg.k_bias, k[i * h ..][0..h]);
            try vmm(io, dev, m, pt, vmat_, norm, vv[i * h ..][0..h], cscratch);
            cfg.addGlue(lg.v_bias, vv[i * h ..][0..h]);
        }
        bidirAttention(q, k, vv, seq, v.num_heads, v.head_dim, scores, attn);
        for (0..seq) |i| {
            try vmm(io, dev, m, pt, omat, attn[i * h ..][0..h], tmp[0..h], cscratch);
            cfg.addGlue(lg.o_bias, tmp[0..h]);
            for (rows[i * h ..][0..h], tmp[0..h]) |*r, o| r.* += o;
        }

        // MLP.
        const fc1m = vmat(cfg, &name_buf, li, "fc1") orelse return error.MissingMatrix;
        const fc2m = vmat(cfg, &name_buf, li, "fc2") orelse return error.MissingMatrix;
        cfg.readGlue(lg.ln2_gamma, gamma);
        cfg.readGlue(lg.ln2_beta, beta);
        for (0..seq) |i| {
            const r = rows[i * h ..][0..h];
            ops.layerNorm(r, gamma, beta, eps, norm);
            try vmm(io, dev, m, pt, fc1m, norm, fc1o, cscratch);
            cfg.addGlue(lg.fc1_bias, fc1o);
            ops.geluInPlace(fc1o);
            try vmm(io, dev, m, pt, fc2m, fc1o, tmp[0..h], cscratch);
            cfg.addGlue(lg.fc2_bias, tmp[0..h]);
            for (r, tmp[0..h]) |*rr, d| rr.* += d;
        }
    }

    // 4. Final LayerNorm.
    cfg.readGlue(v.post_ln_gamma, gamma);
    cfg.readGlue(v.post_ln_beta, beta);
    for (0..seq) |i| {
        const r = rows[i * h ..][0..h];
        ops.layerNorm(r, gamma, beta, eps, norm);
        @memcpy(r, norm);
    }

    // 5. Projector into the text embedding space.
    const l1 = cfg.matrix("projector.linear_1") orelse return error.MissingMatrix;

    if (scale > 1) {
        // Idefics3: pixel-shuffle each output token (space-to-depth merge of a
        // scale x scale block, `h*scale^2` channels) then a single linear. Exact
        // replica of golden/projector.dart pixelShuffle's index mapping.
        const side = std.math.sqrt(seq); // seq is a perfect square (no class token)
        const os = side / scale;
        const es = h * scale; // e*scale
        const out_e = h * scale * scale;
        const shuf = try gpa.alloc(f32, out_e);
        defer gpa.free(shuf);
        for (0..proj_tokens) |o| {
            const h2 = o / os;
            const w2 = o % os;
            for (0..out_e) |c| {
                const hh = c / es;
                const rem = c % es;
                const ww = rem / h;
                const ec = rem % h;
                const in_tok = (h2 * scale + hh) * side + (w2 * scale + ww);
                shuf[c] = rows[in_tok * h + ec];
            }
            const dst = out[o * out_dim ..][0..out_dim];
            try vmm(io, dev, m, pt, l1, shuf, dst, cscratch);
            if (proj.bias1) |b| cfg.addGlue(b, dst);
        }
        return;
    }

    const l2: ?model.Matrix = if (proj.num_layers == 2)
        (cfg.matrix("projector.linear_2") orelse return error.MissingMatrix)
    else
        null;
    for (0..seq) |i| {
        const r = rows[i * h ..][0..h];
        const dst = out[i * out_dim ..][0..out_dim];
        if (l2) |l2m| {
            try vmm(io, dev, m, pt, l1, r, projh, cscratch);
            if (proj.bias1) |b| cfg.addGlue(b, projh);
            ops.geluInPlace(projh);
            try vmm(io, dev, m, pt, l2m, projh, dst, cscratch);
            if (proj.bias2) |b| cfg.addGlue(b, dst);
        } else {
            try vmm(io, dev, m, pt, l1, r, dst, cscratch);
            if (proj.bias1) |b| cfg.addGlue(b, dst);
        }
    }
}
